### ISSUE TITLE
EPIC: implement Javascript & Browser support in Testground

### ISSUE THREAD

**Issue by KatarinaKeti** [testground/testground#1386]
## Description

Testground supports js test written for node, through the `docker:node` builder (legacy) and `docker:generic`.

We want to use Testgroung to test interoperability between different implementations of libp2p.
A critical need is to test browser-related features such as https://github.com/libp2p/specs/pull/412

## What defines this endeavour to be complete?

- [x] Create an `example-browser` plan with an `helloworld` and `failure`  tests (see steps below),
- [x] Create a more complex test case that does a ping between a browser instance and a node instance, using composition files.
- [ ] Finally, the team should be able to reuse the code to implement the browser ping test in https://github.com/libp2p/test-plans

### **Step 1: Run an helloworld test**

(merged in https://github.com/testground/testground/pull/1474)

- [x] 1.1 Succes Test: create a testground PR with a `plans/example-browser` test 
  - Likely will use the code from @GlenDC's [playwright demo](https://github.com/GlenDC/tg-playwright-test/tree/main/plans/playwright-test)
  - [x] It starts the node process, using docker generic, and we call `yarn run testground`
  - [x] We won't use a server, we'll just start the browser, log an HelloWorld, then record a success
- [x] 1.2 Failure test: create another case where we fail on purpose
  - [x] throw an error on the test.
- [x] 1.3 Turn into integration tests
  - [x] add a new test in `./integration_tests/` (feel free to update this ticket with questions)

### **Step 2: Run the helloworld test from WITHIN the browser runtime**

- [x] Use the js-sdk from within the browser to record message, metrics, etc.
- We should end up with a "wrapper" that loads the browser and pass in the environment. Then the test lives in the browser.
- This "should" work, testground-sync service was [rewritten](https://github.com/testground/sync-service/issues/4) a while ago to support connecting from a browser.
- :warning: we'll need to forward the environment variables for the testground SDK to work
- :warning: we might have issues with loading the js-sdk in the browser (missing dist package, etc)

Work threads:

- sdk-js PR: https://github.com/testground/sdk-js/pull/25
- testground PR: https://github.com/testground/testground/pull/1502

### **Step 3: Create a composition with node + js**

- [x] The high level goal is to create a ping-pong test between a node and a browser node.
    - done in https://github.com/testground/testground/pull/1539#issuecomment-1340703249

## Follow-up and other tasks

Discussion and notes raised during the talk about this feature:

- It would make sense to provide a typescript version of the js-sdk
- Think about:
  - Porting tests from https://github.com/little-bear-labs/js-libp2p-webrtc/tree/ckousik/interop-test into the `libp2p/test-plans` repository,
  - implementing the interop tests in node
  - implementing the interop tests in browser

**Comment by brdji** [comment]
I'll write down where I stopped for future reference:

I have managed to write basic [Cypress](https://www.cypress.io/) tests and run them successfully with Testground, _but_ only when running a single instance,
When running tests using multiple instances, if one instance runs the server, other instances cannot connect to it, and Cypress tests fail. Cypress seems to behave strangely when running a Docker container - no logs are written anywhere (`cy.log()`) and errors are difficult to understand (a generic `test failed` message is very common).

What makes this even more strange is that connecting to the server using a simple `GET` request (e.g. using the `axios` library) works, and instances can fetch data from the server that is running on a different instance.

- [ ] Figure out why Cypress cannot connect to a server that is running on a different instance

**Comment by laurentsenta** [comment]
When we give some energy to this ticket:

- We should look into other frameworks. My intuition is that puppeteer is more suited to act as a Testground test node.
- We should connect with potential users, like js-libp2p maintainers (cc @achingbrain) because some features might be critical for them (WebRTC support, etc).

**Comment by mxinden** [comment]
The libp2p project is currently specifying and implementing WebRTC as a libp2p transport. The goal is (a) to enable browsers to connect to nodes that don't possess a TLS certificate within the browsers chain of trust and (b) long term to enable browsers to connect to other browsers.

- Specification https://github.com/libp2p/specs/pull/412
- JS implementation https://github.com/little-bear-labs/js-libp2p-webrtc/
- https://github.com/libp2p/go-libp2p/pull/1655
- Rust implementation https://github.com/libp2p/rust-libp2p/pull/2622

We want to test the libp2p WebRTC transport with Testground. We can already do so with the Go and Rust implementation. Though for the JS implementation we would need some way to run the js-libp2p WebRTC implementation in a browser within a Testground test instance.

---

The @little-bear-labs team (@ckousik and @John-LittleBearLabs) is implementing the js-libp2p WebRTC implementation. At some point they will need to implement the Testground test to test interoperability of the js-libp2p WebRTC implementation with the go- and rust-libp2p implementation. Timeline still unclear.

@laurentsenta in case no one else worked on this issue in the meantime, would you be open for help on this issue from the @little-bear-labs team?

**Comment by ddimaria** [comment]
> I'll write down what I figured out so far:
> 
> I have managed to write basic [Cypress](https://www.cypress.io/) tests and run them successfully with Testground, _but_ only when running a single instance, When running tests using multiple instances, if one instance runs the server, other instances cannot connect to it, and Cypress tests fail. Cypress seems to behave strangely when running a Docker container - no logs are written anywhere (`cy.log()`) and errors are difficult to understand (a generic `test failed` message is very common).
> 
> What makes this even more strange is that connecting to the server using a simple `GET` request (e.g. using the `axios` library) works, and instances can fetch data from the server that is running on a different instance.
> 
> * [ ]  Figure out why Cypress cannot connect to a server that is running on a different instance

@brdji do you have a branch or fork of this experiment?  I'd like to investigate this use case further.  Thanks!

**Comment by achingbrain** [comment]
There's some great research being done here.  I'd like to express quite a strong preference for not using anything Cypress/Selenium-related and instead using something based on [playwright-test](https://www.npmjs.com/package/playwright-test) that lets us run the same test code in node and the browser, runs headlessly, is more lightweight, is easily observable, profilable, etc.

**Comment by GlenDC** [comment]
Hi, I'm new to the discussion here. and part of @little-bear-labs team. Would like to pick up this issue if it is still available. Playwright seems like an excellent choice of me. Can I start working from the master branch of testground or is there some other branch you need me to start from?

@ddimaria do you have that test code available somewhere, so that I can use it as a starting point. If no longer available to you, not an issue, I can also cook something else up I imagine.

**Comment by mxinden** [comment]
:wave: Hi @GlenDC,

> Would like to pick up this issue if it is still available.

It is still available. To the best of my knowledge no one has started working on it yet. @laurentsenta also planned on looking into this, but I am sure he appreciates the additional help. Unfortunately @laurentsenta is out this week.

> Can I start working from the master branch of testground or is there some other branch you need me to start from?

I think the master branch is the best starting point.

> My plan of attack is to have some first "simple" test(s) running with testground over playwright first and I can make a first report after that.

Sounds good to me. Off the top of my head milestones could be:

1. Have a single Testground instance print "Hello World" via playwright (browser) and [sdk-js](https://github.com/testground/sdk-js) and exit with a `0` exit code.
2. Have a Testground instance using playwright (browser) and sdk-js do an HTTP request to a [sdk-go](https://github.com/testground/sdk-go) Testground instance.
3. ...
4. Bring libp2p into the game.

Happy for alternative suggestions.

> We can then decide if we want to continue with Playwright for the remainder of the implementation.

Sounds good to me. In general, I think @achingbrain should be the technical lead on anything JS related.

@GlenDC once you started on this, please post regular updates here to make sure we don't duplicate work.

Great to have you here!

**Comment by brdji** [comment]
> @brdji do you have a branch or fork of this experiment? I'd like to investigate this use case further. Thanks!

I've pushed all code as it is to the following repo, along with instructions on how to run the tests:
https://github.com/testground/tg-cypress-test

>Have a single Testground instance print "Hello World" via playwright (browser) and [sdk-js](https://github.com/testground/sdk-js) and exit with a 0 exit code.
    Have a Testground instance using playwright (browser) and sdk-js do an HTTP request to a [sdk-go](https://github.com/testground/sdk-go) Testground instance.

Using cypress, it is possible to run tests on a single TG instance, but the cypress library cannot connect to an express server running on another TG instance. I suggest the having the following test before attempting any libp2p integration:
1. Have a TG instance run a simple (e.g. express) server, and another instance try to connect to it using Playwright.

**Comment by GlenDC** [comment]
I agree with your plan of attack @mxinden and @brdji and I'll keep you posted on progress. I got through all the documentation yesterday and played around with testground, so you can probably expect a first update this week.

Would it make sense by the way to provide a `docker:playwright` builder by the way, that we can then run using the generic docker runner?

**Comment by brdji** [comment]
I believe our intention is to avoid implementing additional builders, and favor the `generic:docker`. Adding more builders means more dev time must be spent ensuring they are up to date and functional.

**Comment by GlenDC** [comment]
That sounds good. For a first test I was working with `generic:docker` as well. This can definitely work for our purposes, so sounds good. If you want to have different thoughts we can also talk about that, but `docker:generic` does sound reasonable to me, so I'll continue that path.

You can find a first test at <https://github.com/GlenDC/tg-playwright-test/tree/dc73b820241456b992ff47097c24474ad40895a9>. It doesn't do anything fancy yet. Just a single runner which contains an express server as well as a playwright test (headless). All tested using `docker:generic`.

The next step I am investigating is how we would combine things with `compositions`. First goal would to have an express  server running as one group, and a playwright one as a second group. Also hopefully allowing them to communicate via the sync API, to ensure the playwright can wait on the express server being ready. Inspired here by <https://github.com/libp2p/test-plans/blob/master/ping/_compositions/go-rust-interop-latest.toml>.

Will keep you up to date how that goes.

**Comment by mxinden** [comment]
> I believe our intention is to avoid implementing additional builders, and favor the `generic:docker`. Adding more builders means more dev time must be spent ensuring they are up to date and functional.

I think this is the right approach. This is why we have not implemented a `rust:docker`.

**Comment by laurentsenta** [comment]
👍 on all that has been said, welcome @GlenDC! and thanks for helping move this forward @mxinden @brdji,

Happy to help if needed,
this composition might be more palatable to get started: https://github.com/testground/testground/blob/809c50025b7767566f54a14112e3379f2060de24/plans/_integrations_mixed_builders/_compositions/issue-1357-mix-builder-configuration.toml

- you can use `docker:generic` for playwright, and `docker:node` for the express server if that helps
- you can use different paths for the node / browser code

**Comment by GlenDC** [comment]
Nice to meet you @laurentsenta , looking forward to working with you.

I was basing myself on the libp2p testplans, and so far I have this compose file https://github.com/GlenDC/tg-playwright-test/blob/b18f915f8080f97d3f886bc7de3a6872f0e2e5bf/plans/playwright-composition-test/composition.toml:

- it can build, not yet run, but that is probably a minor mistake on my side. I'll look into getting that working first
- for now I was using docker:generic for both, I'll see to use docker:node for the express server, didn't know it existed

I'll let you know in the coming days once I made more progress that is worth sharing!

What I didn't figure out yet is how I can have different plans for the different groups? Because I had in mind that I would for example have one plan to run a server, while another group might really have one or multiple plans for all its test cases. Or is that not how it is to be used? Is it really required that all groups always have the same plan (names)?

Also, what is the best source for me to read documentation and detailed knowledge about these different TOML files. As for now it looks like a black box to me. Is this already well documented or do I need to look at the source code for the time being to figure out such knowledge?

**Comment by GlenDC** [comment]
`docker:node` does not exist, right? I am still using `docker:generic` for now.

**Comment by mxinden** [comment]
> What I didn't figure out yet is how I can have different plans for the different groups? Because I had in mind that I would for example have one plan to run a server, while another group might really have one or multiple plans for all its test cases. Or is that not how it is to be used? Is it really required that all groups always have the same plan (names)?

As far as I know the Testground idiomatic way would be to use the same plan, but pass different flags to the binaries, or decide on the role based on at-runtime coordination via the sync service.

**Comment by GlenDC** [comment]
I now started to read the codebase itself because just using the project as-is with the current documentation is a bit of a though experience:

- The different kind of toml files are not documented, making it hard to write them
- `docker:node` neither is documented, making it also harder to use
- answering questions such as "how do services communicate with one another in testground are also not easy to answer. E.g. what is the IP of the service from which the other one (in a different comp group) can reach it;

Also: the different groups require the same plan, that is not really an issue. However, it is also so that the `case` (test case) is a required global parameter, which thus also has to be shared. It is that that I would have expected not to be shared necessarily.

**Comment by laurentsenta** [comment]
Sorry about that, the documentation is rusty, going through the examples in `plans/` is definitely the way to go at the moment.

I captured your note in #1471 to make sure your comment is not lost.

Some high-level pointers:

- We took the decision to always run a single plan + test case, as a simpler & clearer abstraction. The `path` option acts like a build flag that Testground uses at build time.
- It's fine to pass the group's `role` through a test params, https://github.com/filecoin-project/lotus/blob/7a2bc47486accd1b1f55069b4f1db3781d7710b0/testplans/lotus-soup/_compositions/recovery-k8s.toml#L81
- When it comes to communication, have you had a chance to read the [sync service doc?](https://docs.testground.ai/v/master/concepts-and-architecture/sync-service)

I can review your test and help with making it pass if that's something you need at the moment,

**Comment by GlenDC** [comment]
By the way if you need help in documentation and. the likes, feel free to have issues for them open and assign them to me. I like writing useful documentation, so feel free to do so. The example plans are nice, but I feel that as soon as I try to deviate from it things break down very quickly.

I'll link you a test as soon as I can, but I honestly get stuck over one issue over another. But might be more useful if I start writing these down, so I have more context to give, so will start doing that more carefully.

The service docs I did look at, but the problem is mostly getting things to run, as I feel 100 things can go wrong prior to even ever being able to run a runner and get to something where I can start to worry about the sync service.

**Comment by GlenDC** [comment]
Perfect. I’ll join that slack and get in touch with you later today or tomorrow, depending when you are available.

**Comment by laurentsenta** [comment]
I updated the task's description with the content of today's call (cc @GlenDC @mxinden @galargh),

@GlenDC If you encounter any blocker, especially testground & js-sdk related, don't hesitate to bump this ticket or share Draft PRs.

**Comment by GlenDC** [comment]
Step (1) is already implemented in https://github.com/testground/testground/pull/1474 AFAIK and is ready for review. Regarding Step (2) I see 3 approaches:

- (a) do not embed testground-sdk (JS) in the browser and use playwright "the normal" way, this would mean running small snippets of code by evaluating JS and returning the results to be used from within NodeJS;
- (b) inject testground-sdk (JS) in the browser by evaluating JS code, and effectively injecting the entire library as a single call;
- (c) run a local express server which hosts a blank html page with as only purpose to serve and import (globally) the desired libraries

I know that you do not like (a), so I'll drop that. Personally I think (b) is a big hack (in the ugly way) and stretches the functionality as-is. Not even sure if it would work, even though knowing it is JS I imagine we can make it work if really needed. As such I would suggest to go for (c).

There is no real downside with going for (c). Running an express server is instant, and serving it on localhost should make it instant. On top of that it makes injecting any library painless. E.g. if we want to test libp2p we would serve on that html page both the libp2p lib as well as testground. Going with (c) we can still run our tests as single in-browser scripts, without having potential troubles.

So (c) is the approach I'll probably go for in the week to come, unless someone has any objections. This would be implemented as an advanced example (or however you want to call it). It can be done in the same PR or a follow-up PR, I'll leave that up to you. If no objections I'll probably add it in the same PR.

**Comment by laurentsenta** [comment]
Thanks for sharing these options @GlenDC
(c) sounds great, :thumbsup: to push this way,

I'll review the PR today,
I would merge it as soon as possible: we should not be shy about merging new working features into the edge (master) version. It might encourage other contributors to take a look, test, and even contribute. + smaller PRs mean better, faster, reviews.

**A note regarding (a) which might help with understanding Testground:**

- A Testground test is quite different from a test you might run in Jest or Cypress, where you would be fine with "just" running a few operations and asserting the test results at the end of each sequence.
- A Testground test behaves like a complete, autonomous, application. The Testground SDK is closer to a storage library (like Redis or firebase) than a testing library.

So when it comes to (a), it's less about personal preference and more about risk avoidance: the time we'll save, we might pay it twice when we have to "elevate" the SDK into the application runtime by creating some sort of RPC system.

=> It might be worth having you write a few larger tests like the nodejs libp2p ping example.

**Comment by GlenDC** [comment]
I noticed that I haven't posted an update here in a while. Last week I had not as much time as I had before to work on it, so things were a bit on a slower pace, but since yesterday things are back at the original speed. Here is what happened since last update:

- PR https://github.com/testground/testground/pull/1474 is ready for being merged, once all required approvals have happened;
- The previous PR adds a first simple playwright example showing you how to run it;

In the meanwhile I started on step (2) in a new branch, for which I'll make a PR once the previous PR is merged in. With this step:

- I embed testground (sdk) in the playwright-controlled browser;
- do I run the tests directly as they are in the browser, where you'll have access to _testground_ via `window.testground`, e.g. `window.testground.runenv`;
   - This is however a bit long, so I'm going to instead change it to requesting a callback in the form of `function(runenv[, client])` such that we can use `runenv` directly instead of via the `window.testground way`.
   - The latter approach will also make it possible to actually have it as a regular script, as for now I was passing them as string-scripts.

For this to work I also made some changes already to testground-sdk (js), for which I'll make a PR as well, once I get it all to a working state. I changed 2 things so far, but only 1 is required. That one thing is:

- prevent the logger from trying to use `fs` (so make the `stdout` file logger opt-out);
  - this is just a hack I think, so we might want to think carefully how we want this API in function of the purpose of the different transports used in the logger;

I'll post later this week once I have first Draft PR's ready, and only once the first PR is merged in.

**Comment by laurentsenta** [comment]
Understood, we're lacking documentation on how to debug a `docker:generic` test plan (it depends a lot on the language).

"Horrible iteration speed" is definitely worth fixing :) Feel free to work on improving this or reach out for a peer programming session.

My experience with local dev on rust has been "slow-ish". Here are a few tips,
if you try them, we can turn those into documentation and eventually automations:


### Are you uploading a large test to the daemon?

After a few iterations, most services should be ready, and the docker layers should be used. So the main overhead caused by Testground is the test upload.

We don't have `.testgroundignore` [yet](https://github.com/testground/testground/pull/1441), so you might be better off removing your node_modules folder (`du -hs ./my-test` to have an idea of the data you're uploading).


### Are you relying on Docker caching?

We talked about the `.dockerignore` and how it's useful to skip files like the `Dockerfile` itself to save rebuilds.

With Docker + node, I usually split the `npm install` from `npm build`.
Because packages take a lot of time to download and don't change often, it's very useful to use Docker caching. Here is an example:

```Dockerfile
COPY package.json package-lock.json* ./
RUN npm ci && npm cache clean --force

COPY ./src ./src
RUN npm run build
```

### Is this a common practice / is it very possible, to be able to run a plan directly from the host machine, just as well as in docker?

Yes, it is possible, at least up to a point where the test WILL run in Testground. (Testground is "just" starting a bunch of network containers and configuring networks for you).

**1. I would make sure the test can build on my machine (no testground, no docker).**

`npm build`, even maybe `docker build .`

**2. I would make sure the test can run on my machine (no testground, no docker).**

First you need Testground's env variables ([examples](https://github.com/testground/sdk-js/blob/6a9f2138ce6e28fc25eafa6a9b23b0e41bb0cee5/src/runtime/params.js#L7-L21)).

An easy way to get those: run your test once in Testground + Docker and `console.log(process.env)`.
Then export these variables in the env you'll use to run the test.
You'll want to tweak some variables, like `SYNC_SERVICE_HOST="127.0.0.1"` (testground started it for you).

`source ./my-env && npm run`


That would get rid of any Testground overhead.

Note later we might want to investigate Nix builders (https://github.com/testground/testground/pull/1425/files)

**Comment by GlenDC** [comment]
@laurentsenta we might indeed want to pair program this week once or twice, as I feel it is still mostly my struggles with testground that stop me the most. Will reach out to you in the upcoming week.

**Comment by GlenDC** [comment]
Seems there is a bug in the `docker build` phase:

```
task of type build cancelled: docker build failed: COPY failed: file not found in build context or excluded by .dockerignore: stat package.json: file does not exist
```

That is when I try to change the browser example's docker file to something like:

```
RUN mkdir -p /plan
WORKDIR /plan

COPY package.json package-lock.json ./
RUN npm ci && npm cache clean --force
```

which fails on the package.json line. Weird thing is that whn trying to build the Dockerfile myself using `docker build`, it does work.

**Comment by GlenDC** [comment]
Small update on this issue.

- @laurentsenta helped me out with some pair programming, so I am no longer stuck
- we established the following milestones to be reached (now that this first part is shipped to master):
   - 0. improve dockerfile of first example (just to have this out of the way) to improve iteration speed (will be done as part of next PR, not something separate)
   - 1. get our advanced browser example plan in master (also part of next PR, more about this in a bit)
   - 2. add a `js` group plan as part of https://github.com/libp2p/test-plans/tree/master/ping (as part of `/js`) which will stil use docker generic and use the approach paved and showed in (1). For now just latest version of libp2p (js)
   - 3. next steps (that can happen in a yet to be defined order, or in a parallel):
       - (a) one next step will be to supper all relevant last versions of libp2p (js)
       - (b) another next step will be to start working on integrating our browser solution as some native builder/runner, so that we can easily allow people's test plan to work for both the `node` runtime as well as the `browser` runtime;
 
We can consier this issue as "done" once (2) is finishd:

- (3a) is libp2p specific and can be tracked further in the libp2p/test-plans repo as a new (linked) issue;
- (3b) goes beyond the scope of this issue and can be tracked as a new testground issue (linked to this issue);

---

For (1):

- Up to today I was locally developing this as some kind of hybrid solution where we run a browser, spawned from within a node test
- this approach was very clucnky however, and it also meant that our testplans are really more complex than they are and on top of that browser-only
- instead we'll opt to make a "magic" layer which allows you to run your node (js) test plan folder as-is, and have it run both in a node environment as well as a browser environment;
- for the latter (browser env) we will still puppeteer in the background everything from the node (playwright) process, but that will be invisible to the user of testground, and this approach will also be clear already from the advanced browser example, a step and approach pretty different from the hello browser example;

@laurentsenta also drew a very nice diagram for this, which is now acting as my muse (and might not make sense to people outside of me and Laurent):

![image](https://user-images.githubusercontent.com/3900482/195193755-9f158496-69ca-4019-b9f8-77f084b813b4.png)

**Comment by GlenDC** [comment]
@laurentsenta I have a Draft PR ready at https://github.com/testground/testground/issues/1386 for the advanced example testplan. There's still some work, but our desired workflow has been implemented as a PoC and it does seem to work, only step is to actually make the test still run, but we do already have the `testground-sdk` working in the browser and communicating with our other docker containers.

I'm ready to receive feedback and when you have time we can also sit together for another round of pair programming if you wish (or rather exchange ideas and synchronise about a thing or two.

Here are some screenshots of the remote debugger in action (chrome only):

<img width="1550" alt="Screenshot 2022-10-25 at 00 51 27" src="https://user-images.githubusercontent.com/3900482/197645019-35f2b08b-f2b4-465f-bf29-2e794d3a32cc.png">
<img width="641" alt="Screenshot 2022-10-25 at 00 50 02" src="https://user-images.githubusercontent.com/3900482/197645024-a6be23f3-0455-4af5-889b-d883fb5b6518.png">
<img width="1670" alt="Screenshot 2022-10-25 at 00 49 52" src="https://user-images.githubusercontent.com/3900482/197645025-733968c3-edd8-48d2-9d5d-a9fc251ca253.png">

**Comment by GlenDC** [comment]
@laurentsenta https://github.com/testground/testground/pull/1502
success/failure tests now work on node, as well as on firefox / webkit

chrome (chromium) is still hanging on the sync service, so I need to further debug that

Beyond that I will want to:

- sit together with you for a session
- improve the user experience:
  - halting browser prior to exit is now implemented in a very hacked fashion, would want a cleaner solution ideally (to be designed);
  - firefox remote debugging is not working yet, chrome debugging does work, if firefox debugging is also desired it will need to be made working as well;
  - make it somehow possible to pass browser kind via test case arg (right now this is a build arg)
- make the sync test work (fails for the current node version as well in my branch…

And of course:
- https://github.com/testground/sdk-js/pull/25 can use plenty of improvement, but for that we better just sit together :slightly_smiling_face:

**Comment by GlenDC** [comment]
I've created 2 new test ground issues, as these are out of scope for this issue, but are never the less still relevant and very related:

1. https://github.com/testground/testground/issues/1505: currently we cannot expose args such as what browser to use as "runner" args, as the `local:docker` runner doesn't allow variabls to be defined. The goal of this issue is to be able to define environment variables as an overwrite of the default.
2. https://github.com/testground/testground/issues/1506 (this one is not important, but a nice to have): provide the ability to deterministically choose the port on the machine side instead of getting a random free one.

**Comment by GlenDC** [comment]
Update on this issue.

- browser support now works and is finished in PR https://github.com/testground/testground/pull/1502
   - this PR is ready for a final review. It is blocked however until we also get https://github.com/testground/testground/pull/1507 merged in and applied (to allow us to test in the integration plan on firefox, chromium and webkit, without having to build 3 times)
 - `@testground/sdk` (js sdk for testground) required some serious work to get it to work as well as on NodeJS and Browser, that PR is ready for a final review as well: https://github.com/testground/sdk-js/pull/26

The example plan in the final PR (https://github.com/testground/testground/pull/1502) proves that:

- you can have a single JS testplan which can be tested as-is (without any work on the testplan's side) on:
   - NodeJS (`docker:node`)
   - Cross-Browser: Firefox, WebKit and Chromium (`docker:generic`)
     - current test plan tests in the integration test only with chromium, but I did confirm that it works in NodeJS, Firefox, Chromium and Webkit alike, without any modifications required;
   - with the magic layer in between being in the `browser` folder
      - it is this folder and other possible improvements that we could handle automatically as part of some new kind of `docker:browser` builder target, should we wish :)

Things that are considered out of scope for this issue and related PR, but that will follow out of this issue and which I might be able to tackle now that this work is done:
- https://github.com/testground/testground/issues/1506: this would simplify remotely debugging the browser, as currently that port on the host-side (where you would debug from) is randomly chosen from the available ports, making it a bit of a hassle to select which port to remotely debug from your local (host) chrome;
- https://github.com/testground/sdk-js/issues/27: inspired by the `rust-sdk` we could add using testground, a copy of this example plan to test `@testground/sdk` both in browser and in node
- a first `ping` test using testground in the `libp2p/test-plans` (related to https://github.com/libp2p/test-plans/issues/53?)
- `js-libp2p` tests using testground (e.g. https://github.com/libp2p/js-libp2p/issues/1478)
- https://github.com/testground/testground/issues/1521: time improvement for CI runs, easy but non-critica

**Comment by GlenDC** [comment]
While I'm waiting for all my testground and `@tesground/sdk` (sdk-js) PR's to be further reviewed and/or merged, I started dabbing around with ping tests for `libp2p-js`. I will however need to study a bit and learn more about libp2p before I can comfortable continue that PR.

The draft PR for that can be followed at https://github.com/libp2p/test-plans/pull/74 for those who are interested.

**Comment by laurentsenta** [comment]
Awesome, thank you for sharing, I answered on the different PRs.
Just a note:

> this PR is ready for a final review. It is blocked however until we also get https://github.com/testground/testground/pull/1507 merged in and applied (to allow us to test in the integration plan on firefox, chromium and webkit, without having to build 3 times)

That feature will deserve some more time to be implemented completely and covered with testing, since it's an optimization, I'd prefer if we don't block on it and stick to 3 builds (docker layer caching will make these instants anyway).

**Comment by laurentsenta** [comment]
Merging:
https://github.com/testground/testground/pull/1502

A few notes for follow-ups in:
https://github.com/testground/testground/pull/1502#pullrequestreview-1199287441

**Comment by GlenDC** [comment]
Implemented the follow-up notes in my power (AFAIK) as mentioned in #1502 as a new PR: https://github.com/testground/testground/pull/1539

Ready for review.

**Comment by GlenDC** [comment]
PR #1539 has been updated and a composition has been added, ensuring the sync test works cross-runtime, using an instance of firefox, chromium, webkit and node.

**Comment by GlenDC** [comment]
All todo's and comments of https://github.com/testground/testground/pull/1539 have been addressed and it is ready to be merged in today AFAIK.

My son will be born today, so the continuation of https://github.com/libp2p/test-plans/issues/61#issuecomment-1340727625 will have to wait a bit, but I should be able to continue it later this week.

**Comment by p-shahi** [comment]
closing in favor of https://github.com/libp2p/test-plans/pull/104
